hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/lib/MutableRatesWithAggregation.java

```diff
@@ -71,11 +74,20 @@ public void init(Class<?> protocol) {
     if (protocolCache.contains(protocol)) {
       return;
     }
-    protocolCache.add(protocol);
-    for (Method method : protocol.getDeclaredMethods()) {
-      String name = method.getName();
-      LOG.debug(name);
-      addMetricIfNotExists(name);
+
+    List<Class<?>> protocols = new ArrayList<>();
+    protocols.add(protocol);
+    if (protocol.getDeclaredMethods().length == 0) {
+      protocols.addAll(Arrays.asList(protocol.getInterfaces()));
+    }
+
+    for (Class<?> pClass : protocols) {
+      protocolCache.add(pClass);
+      for (Method method : pClass.getDeclaredMethods()) {
+        String name = method.getName();
+        LOG.debug(name);
+        addMetricIfNotExists(name);
+      }
```

Comment:
Seems we can fix this much more simply by just using [getMethods()](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getMethods--) which includes inherited methods. The main difference is that we'll only get public methods, not private/package-private, but I don't think we have any use cases which need those methods anyway. This is focused on protocol classes for which we only need the public methods.

```suggestion
    protocolCache.add(protocol);
    for (Method method : protocol.getMethods()) {
      String name = method.getName();
      LOG.debug(name);
      addMetricIfNotExists(name);
```

Reply:
Thanks for your advice. This method is really more suitable.

Reply:
If we pass XXXProtocol or XXXTranslatorPB, everything will be ok. However, it seems more complex to change related code in RPC server. So I use getMethods() finally.
I enriched the content of testNNRpcMetricsWithNonHA(). When the MiniDFSCluster starts, datanode will call versionRequest(), so the metrics of all methods in DatanodeProtocol should be initialized, including CommitBlockSynchronizationNumOps. Please give a check, thanks.